File: src/db/repositories/project.ts
import { getDb } from "../client.js";
import type { NewProject, Project, ProjectUpdate } from "../schema.js";

export const projectRepo = {
  async findActiveByChat(chatId: number): Promise<Project | undefined> {
    return getDb()
      .selectFrom("projects")
      .selectAll()
      .where("telegram_chat_id", "=", chatId)
      .where("is_active", "=", true)
      .executeTakeFirst();
  },

  async findAllByChat(chatId: number): Promise<Project[]> {
    return getDb()
      .selectFrom("projects")
      .selectAll()
      .where("telegram_chat_id", "=", chatId)
      .orderBy("created_at", "desc")
      .execute();
  },

  async findById(id: number): Promise<Project | undefined> {
    return getDb()
      .selectFrom("projects")
      .selectAll()
      .where("id", "=", id)
      .executeTakeFirst();
  },

  async create(project: NewProject): Promise<Project> {
    return getDb()
      .insertInto("projects")
      .values(project)
      .returningAll()
      .executeTakeFirstOrThrow();
  },

  async update(id: number, update: ProjectUpdate): Promise<Project | undefined> {
    return getDb()
      .updateTable("projects")
      .set(update)
      .where("id", "=", id)
      .returningAll()
      .executeTakeFirst();
  },

  async setActive(chatId: number, projectId: number): Promise<void> {
    const db = getDb();
    // Deactivate all projects for this chat
    await db
      .updateTable("projects")
      .set({ is_active: false })
      .where("telegram_chat_id", "=", chatId)
      .execute();
    // Activate selected
    await db
      .updateTable("projects")
      .set({ is_active: true })
      .where("id", "=", projectId)
      .execute();
  },

  async deleteById(id: number): Promise<void> {
    await getDb().deleteFrom("projects").where("id", "=", id).execute();
  },
};

File: proxima/db/repositories/project.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from sqlalchemy import desc, select, update

from proxima.db.engine import Database
from proxima.db.models import Project


class ProjectRepository:
    def __init__(self, db: Database) -> None:
        self.db = db

    async def find_active_by_chat(self, chat_id: int) -> Project | None:
        async with self.db.session() as session:
            result = await session.execute(
                select(Project)
                .where(Project.telegram_chat_id == chat_id)
                .where(Project.is_active.is_(True))
                .limit(1)
            )
            return result.scalar_one_or_none()

    async def find_all(self) -> list[Project]:
        async with self.db.session() as session:
            result = await session.execute(select(Project))
            return list(result.scalars().all())

    async def find_all_by_chat(self, chat_id: int) -> list[Project]:
        async with self.db.session() as session:
            result = await session.execute(
                select(Project)
                .where(Project.telegram_chat_id == chat_id)
                .order_by(desc(Project.created_at))
            )
            return list(result.scalars().all())

    async def find_by_id(self, project_id: int) -> Project | None:
        async with self.db.session() as session:
            result = await session.execute(select(Project).where(Project.id == project_id).limit(1))
            return result.scalar_one_or_none()

    async def create(self, values: Mapping[str, Any]) -> Project:
        project = Project(**dict(values))
        async with self.db.session() as session:
            session.add(project)
            await session.commit()
            await session.refresh(project)
            return project

    async def update(self, project_id: int, values: Mapping[str, Any]) -> Project | None:
        async with self.db.session() as session:
            result = await session.execute(select(Project).where(Project.id == project_id).limit(1))
            project = result.scalar_one_or_none()
            if not project:
                return None
            for key, value in values.items():
                setattr(project, key, value)
            await session.commit()
            await session.refresh(project)
            return project

    async def set_active(self, chat_id: int, project_id: int) -> None:
        async with self.db.session() as session:
            await session.execute(
                update(Project).where(Project.telegram_chat_id == chat_id).values(is_active=False)
            )
            await session.execute(
                update(Project).where(Project.id == project_id).values(is_active=True)
            )
            await session.commit()

    async def delete_by_id(self, project_id: int) -> None:
        async with self.db.session() as session:
            project = await session.get(Project, project_id)
            if project is None:
                return
            await session.delete(project)
            await session.commit()
